bundles/org.eclipse.cdt.lsp/src/org/eclipse/cdt/lsp/ui/navigator/CSymbolsContentProvider.java

```diff
@@ -115,18 +160,20 @@ protected void refreshTreeContentFromLS(IFile file) {
 					}
 				}
 				symbols.thenAcceptAsync(response -> {
-					symbolsModel.setUri(file.getLocationURI());
-					symbolsModel.update(response);
+					thisSymbolsContainer.symbolsModel.setUri(file.getLocationURI());
+					thisSymbolsContainer.symbolsModel.update(response);
+					//update/create container:
+					cachedSymbols.put(file.getLocationURI(), thisSymbolsContainer);
 				}).join();
-			} else {
-				temporaryLoadedDocument = false;
-				symbolsModel.setUri(file.getLocationURI());
-				symbolsModel.update(null);
 			}
 		} catch (Exception e) {
 			Platform.getLog(getClass()).error(e.getMessage(), e);
 		} finally {
-			if (temporaryLoadedDocument) {
+			// disconnect file if it's shared by somebody else to ensure
```

Comment:
The logic here doesn't seem correct to me - if we aren't connecting (i.e. adding to the reference count) we shouldn't be disconnecting (reducing reference count). The call to LSPEclipseUtils.getDocument when  getExistingDocument == null that was here before did the connect, hence why you needed temporaryLoadedDocument to know if you needed to do a disconnect. 

With the update above where LSPEclipseUtils.getDocument(file) is called unconditionally it looks like we are doing a connect sometimes, but it may not be shared and the handle we have to the buffer is not being kept track of.

Reply:
Agree. I did it that way because I wanted to get notified when the last document has been disconnected. The idea was to remove the cached symbols from the HashMap when the corresponding documents getting disconnected. 

In the new patch set the cached symbols won't be removed, because I cannot do it in the `bufferDisposed` event anymore: When the file is not opened in the editor and the temporary loaded document gets disconnected in the finally stage, `bufferDisposed` gets called and deletes the cached symbols.